[dynamo] Config to not emit runtime asserts

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122406

Introducing dynamo config.do_not_emit_runtime_asserts (which can be set with env var TORCH_DYNAMO_DO_NOT_EMIT_RUNTIME_ASSERTS=1)

The main idea is to be able to remove runtime asserts from the final graph, added for symbolic shape hints.
The intention is coming from asserts in torchrec, where we have shapes as tensor elements on cuda and have to hint on tensor items. This could be costly for perf and this will be the way how to test the effect of asserts and be able to remove them.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

Differential Revision: [D55192671](https://our.internmc.facebook.com/intern/diff/D55192671)

IvanKobzarev committed:
[dynamo] Config to not emit runtime asserts

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55192671).

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -1092,6 +1094,9 @@ def compile_and_call_fx_graph(self, tx, rv, root):
         self.real_value_cache.clear()
 
+        if config.do_not_emit_runtime_asserts:
+            self.graph.eliminate_dead_code()
```
This seems suspicious, because the graph here can have mutations in it, I don't think this is actually safe.

IvanKobzarev committed:
Update on "[dynamo] Config to not emit runtime asserts"


Introducing dynamo config.do_not_emit_runtime_asserts (which can be set with env var TORCH_DYNAMO_DO_NOT_EMIT_RUNTIME_ASSERTS=1)

The main idea is to be able to remove runtime asserts from the final graph, added for symbolic shape hints.
The intention is coming from asserts in torchrec, where we have shapes as tensor elements on cuda and have to hint on tensor items. This could be costly for perf and this will be the way how to test the effect of asserts and be able to remove them.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D55192671](https://our.internmc.facebook.com/intern/diff/D55192671)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[dynamo] Config to not emit runtime asserts"


Introducing dynamo config.do_not_emit_runtime_asserts (which can be set with env var TORCH_DYNAMO_DO_NOT_EMIT_RUNTIME_ASSERTS=1)

The main idea is to be able to remove runtime asserts from the final graph, added for symbolic shape hints.
The intention is coming from asserts in torchrec, where we have shapes as tensor elements on cuda and have to hint on tensor items. This could be costly for perf and this will be the way how to test the effect of asserts and be able to remove them.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D55192671](https://our.internmc.facebook.com/intern/diff/D55192671)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55192671).
